lib/saml/auth_fail_handler.rb

```diff
@@ -17,15 +19,17 @@ def initialize(saml_response, user, session)
       @context = nil
     end
 
-    def known_error?
-      known_errors = [
-        :clicked_deny?,
-        :auth_too_late?,
-        :auth_too_early?
-      ]
+    def error
+      KNOWN_ERRORS.each do |known_error|
+        return known_error if send("#{known_error}?")
+      end
```

Comment:
So this gets you O(1) vs O(n) complexity but also more readable and more DRY because the logic for 1 or more errors is handled in one place, rather than in each of individual private methods.